src/storage/v2/schema_validator.cpp

```diff
@@ -0,0 +1,107 @@
+
+[[nodiscard]] std::optional<SchemaViolation> SchemaValidator::ValidateVertexCreate(
+    LabelId primary_label, const std::vector<LabelId> &labels,
+    const std::vector<std::pair<PropertyId, PropertyValue>> &properties) const {
+  // Schema on primary label
+  const auto schema = schemas_.GetSchema(primary_label);
+  if (!schema) {
+    return SchemaViolation(SchemaViolation::ValidationStatus::NO_SCHEMA_DEFINED_FOR_LABEL, primary_label);
+  }
+
+  // Is there another primary label among secondary labels
+  for (const auto &secondary_label : labels) {
+    if (schemas_.GetSchema(secondary_label) != std::nullopt) {
+      return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_SECONDARY_LABEL_IS_PRIMARY, secondary_label);
+    }
+  }
+
+  // Check only properties defined by schema
+  for (const auto &schema_type : schema->second) {
+    // Check schema property existence
+    auto property_pair = std::ranges::find_if(
+        properties, [schema_property_id = schema_type.property_id](const auto &property_type_value) {
+          return property_type_value.first == schema_property_id;
+        });
+    if (property_pair == properties.end()) {
+      return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_HAS_NO_PROPERTY, primary_label, schema_type);
+    }
+
+    // Check schema property type
+    if (auto property_schema_type = PropertyTypeToSchemaType(property_pair->second);
+        property_schema_type && *property_schema_type != schema_type.type) {
+      return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_PROPERTY_WRONG_TYPE, primary_label, schema_type,
+                             property_pair->second);
+    }
+  }
+
+  return std::nullopt;
+}
+
+[[nodiscard]] std::optional<SchemaViolation> SchemaValidator::ValidateVertexUpdate(const LabelId primary_label,
+                                                                                   const PropertyId property_id) const {
+  // Verify existence of schema on primary label
+  const auto schema = schemas_.GetSchema(primary_label);
+  if (!schema) {
+    return SchemaViolation(SchemaViolation::ValidationStatus::NO_SCHEMA_DEFINED_FOR_LABEL, primary_label);
+  }
+
+  // Verify that updating property is not part of schema
+  if (const auto schema_property = std::ranges::find_if(
+          schema->second,
+          [property_id](const auto &schema_property) { return property_id == schema_property.property_id; });
+      schema_property != schema->second.end()) {
+    return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_UPDATE_PRIMARY_KEY, primary_label,
+                           *schema_property);
+  }
+  return std::nullopt;
+}
+
+[[nodiscard]] std::optional<SchemaViolation> SchemaValidator::ValidateLabelUpdate(const LabelId label) const {
+  const auto schema = schemas_.GetSchema(label);
+  if (schema) {
+    return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_ALREADY_HAS_PRIMARY_LABEL, label);
```

Comment:
```suggestion
    return SchemaViolation(SchemaViolation::ValidationStatus::VERTEX_MODIFY_PRIMARY_LABEL, label);
```

The name `VERTEX_ALREADY_HAS_PRIMARY_LABEL` suggest for me that a vertex might exist without a primary label, which shouldn't be allowed in my opinion. Also this function is also called from `RemoveLabel`, which doesn't want to add any new label.